tower/src/balance/p2c/service.rs

```diff
@@ -18,24 +18,15 @@ use tokio::sync::oneshot;
 use tower_service::Service;
 use tracing::{debug, trace};
 
-/// Distributes requests across inner services using the [Power of Two Choices][p2c].
+/// Efficiently distributes requests across an arbitrary number of services.
 ///
-/// As described in the [Finagle Guide][finagle]:
-///
-/// > The algorithm randomly picks two services from the set of ready endpoints and
-/// > selects the least loaded of the two. By repeatedly using this strategy, we can
-/// > expect a manageable upper bound on the maximum load of any server.
-/// >
-/// > The maximum load variance between any two servers is bound by `ln(ln(n))` where
-/// > `n` is the number of servers in the cluster.
+/// See the [module-level documentation](..) for details.
 ///
 /// Note that `Balance` requires that the `Discover` you use is `Unpin` in order to implement
 /// `Service`. This is because it needs to be accessed from `Service::poll_ready`, which takes
 /// `&mut self`. You can achieve this easily by wrapping your `Discover` in [`Box::pin`] before you
 /// construct the `Balance` instance. For more details, see [#319].
```

Comment:
probably best saved for a follow-up PR, but should `Discover` just have an `Unpin` bound on it? Since the unpin bound is only on the `Service` impl for `Balance`, this will currently fail _very_ late, when a user tries to wrap a `Balance` in another layer or actually call a `Balance`, and it probably won't be super clear why `Service` isn't implemented. If `Discover` required `Self: Unpin`, the compiler error would occur when the user writes a bad `Discover` impl, which would be much clearer.

Reply:
I think we should avoid `Unpin` bounds _wherever possible_. They are really sad, especially when we one day get generators that implement `Stream`. This also ties into the larger discussion in #319.

Reply:
Hmm, but if `Discover` is going to be polled in `poll_ready`, it _will_ always have to be `Unpin` unless `poll_ready` is changed to take a `Pin<&mut Self>` receiver. In that case, we would be making a breaking change anyway and could remove an `Unpin` bound from `Discover`.

Reply:
I don't know if `Discover` will only ever be polled in `poll_ready` though? But you're right that if `poll_ready` later changes to be `Pin`, we could remove the bound then (although I think _removing_ the bound is backwards-compatible anyway?).